src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -472,22 +472,70 @@ entails SolverOptions{..} constraint context hints =
+            -- If the first argument is a newtype applied to some other types
+            -- (e.g. @newtype T a = T a@ in @Coercible (T X) b@), look up the
+            -- type of its wrapped field and yield a new wanted constraint in
+            -- terms of that type with the type arguments substituted in (e.g.
+            -- @Coercible (T[X/a]) b = Coercible X b@ in the example).
+            let wrappedTySub = replaceAllTypeVars (zip tvs xs) wrappedTy
+            pure $ Just [srcCoercibleConstraint kindType wrappedTySub b]
+          _ -> pure Nothing
       | otherwise =
         -- In all other cases we can't solve the constraint.
         pure Nothing
 
     srcCoercibleConstraint :: SourceType -> SourceType -> SourceType -> SourceConstraint
     srcCoercibleConstraint k a b = srcConstraint C.Coercible [k] [a, b] Nothing
 
+    -- | Looks up a given name and, if it names a newtype, returns the names of the
+    -- type's parameters, the type the newtype wraps and the names of the type's
+    -- fields.
+    lookupNewtypeConstructor
+      :: Maybe (ModuleName, [(SourceAnn, ModuleName, ImportDeclarationType, Maybe ModuleName)])
+      -> Environment
+      -> Qualified (ProperName 'TypeName)
+      -> Maybe ([Text], SourceType, [Ident])
+    lookupNewtypeConstructor currentModule env qualifiedNewtypeName@(Qualified newtypeModuleName newtypeName) = do
+      let currentModuleName = fst <$> currentModule
+          newtypeImportsDeclType = foldMap (foldMap toNewtypeImportDeclType . snd) currentModule
+      guard $ newtypeModuleName == currentModuleName || isNewtypeCtorInScope newtypeImportsDeclType
+      (_, DataType tvs [(ctorName, [wrappedTy])]) <- M.lookup qualifiedNewtypeName (types env)
+      (Newtype, _, _, ids) <- M.lookup (Qualified newtypeModuleName ctorName) (dataConstructors env)
+      pure (map (\(name, _, _) -> name) tvs, wrappedTy, ids)
+      where
+      toNewtypeImportDeclType (_, importedModuleName, importDeclType, _)
+        | Just importedModuleName == newtypeModuleName = pure importDeclType
+        | otherwise = empty
+      isNewtypeCtorInScope = maybe False $ \case
+        Implicit -> True
+        Explicit refs | any isNewtypeCtorRef refs -> True
+        Hiding refs | not $ any isNewtypeCtorRef refs -> True
+        _ -> False
+      isNewtypeCtorRef = \case
+        TypeRef _ importedTyName Nothing -> importedTyName == newtypeName
+        TypeRef _ importedTyName (Just [_]) -> importedTyName == newtypeName
+        _ -> False
+
+    isUnusedNewtypeCtorImportWarning
+      :: Maybe ModuleName
+      -> Qualified (ProperName 'TypeName)
+      -> ErrorMessage
+      -> Bool
+    isUnusedNewtypeCtorImportWarning currentModuleName qualifiedNewtypeName = \case
+      ErrorMessage hints' (UnusedDctorImport unusedTyModuleName unusedTyName _ _) ->
+        any isInCurrentModule hints' && qualifiedNewtypeName == mkQualified unusedTyName unusedTyModuleName
+      _ -> False
```

Comment:
What about `UnusedImport`s? Suppose you have
```purs
module N1 where

newtype N1 a = N1 a
```
```purs
module N2 where

import N1 (N1)

newtype N2 a = N2 (N1 a)
```
```purs
module Unwrap where

import Safe.Coerce (coerce)
import N1
import N2 (N2(..))

unwrap :: forall a. N2 a -> a
unwrap = coerce
```

That ought not to warn about the `import N1`, right? But I think it will.

Reply:
Good catch! I added a case and test.